### PR TITLE
759 3x tripal registration form

### PR DESCRIPTION
**Pull request by shawnawsu**
# New Feature

Issue #759 

## Description

The form will be an opt-in registration so users can choose not to register the site if it's a dev site or if they simply do not wish to share that they are using Tripal. Users will be reminded to register the site until they either do so or check the 'disable reminders' checkbox.

This update will allow us to push out notifications about Tripal and Tripal Extension module updates that are ready for download. Additionally, it will provide more accurate usage statistics since many users are not downloading the module through drupal.org.

## Testing?

Pull the branch and clear your cache, no update.php required. 
You should now have a drupal message asking you to register your site.
Navigate to the tripal menu (admin/tripal) and ensure the Registration menu link is there. 
Click on the registration tab to get to the form. 
Fill out the form and submit it. 
Once submitted check tripal.info to ensure the registration was successfully received. 
Also, check that the registration reminder message has disappeared.

### CODE REVIEW

File: tripal/includes/tripal.registration.inc [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+      '#type' => 'item',
+      '#markup' => t('Registering your site is important for continued improvements to the software. Tripal is open-source
+          and dependent on external funding. When you register your site, it provides important details that can
+          be used to justify continued support from funding agencies. You may opt-in by providing answers to the 
+          following questions or opt-out by checking the box below. If you opt-in, your site will periodically
+          connect to the http://tripal.info website and provide updated registration details. If you opt-out, no
+          registration information is communictated.'),
+    ];
+
+  $form['disable_tripal_reporting'] = array(
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc:41]: I should just be able to press the opt-out box and submit.  However, there are validation errors if i do that:

Screen Shot 2019-04-01 at 3 45 11 PM

I would propose we skip validation if we're opting out.

File: tripal/includes/tripal.registration.inc [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+      '#required' => TRUE,
+      '#description' => t('Please register your site regardless if it is experimental (to explore tripal), 
+       for development of a future site (or updates to an existing site), or a site currently 
+       in production. For funding, it is important to know how many sites are active for each category.  If your site changes 
+       status, such as from development to production, please remember to return and update the purpose.')
+  );
+
+  $form['details']['tripal_reg_site_modules'] = array(
+      '#type' => 'checkbox',
+      '#default_value' => isset($form_data['values']['tripal_reg_site_modules']) ? $form_data['values']['tripal_reg_site_modules'] : NULL,
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc]: Can we please have this  checked by default?

```suggestion
      '#default_value' => isset($form_data['values']['tripal_reg_site_modules']) ? $form_data['values']['tripal_reg_site_modules'] : 1,
```

File: tripal/includes/tripal.registration.inc [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+ * @param $form
+ *   A structured array containing the elements and properties of the form.
+ * @param $form_state
+ *   An array that stores information about the form's current state 
+ *   during processing.
+ */
+function tripal_registration_form_validate($form, &$form_state){
+  $mail_pi = $form_state['values']['principal_investigator_email'];
+  $mail_sa = $form_state['values']['tripal_reg_site_admin_email'];
+  if (!valid_email_address($mail_pi)) {
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc]: I get this error when submitting  no PI email.  However the wording implies its not a required field.  If its required, there needs to be a way to opt out of email notices.  If it isnt required, then dont run `valid_email_address` if the value is null

Screen Shot 2019-04-01 at 3 45 11 PM

File: tripal/includes/tripal.registration.inc [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    '#type' => 'textfield',
+    '#title' => t('Email'),
+    '#default_value' => isset($form_data['values']['principal_investigator_email']) ? $form_data['values']['principal_investigator_email'] : NULL,
+    '#required' => FALSE,
+  );
+  $form['details']['tripal_reg_site_admin'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Site Manager (if different from the principal investigator)'),
+    '#collapsible' => TRUE,
+    '#collapsed' => FALSE,
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc]: I suggest collapsing this.  The form is a bit long for optional info.  by collapsing non-essential stuff we'll make it look less terrible?


edit:  after going through the form, i think having the 3 funding sources expanded makes it look longer than it is.   If we only start with 1 funding source open perhaps this suggestion is less necessary.  Very weird, my first time open it had 3 funding sources, but now it is only showing the 1...

```suggestion
    '#collapsed' => TRUE,
```

File: tripal/includes/tripal.registration.inc [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+  $count = count($form_data['values']['funding']);
+  $form_state['details']['funding'] = $form_state['details']['funding'] + $count;
+  for ($i = 1; $i <= $form_state['details']['funding']; $i++) {
+    $form['details']['funding'][$i] = array(
+      '#type' => 'fieldset',
+      '#title' => t("Funding Source $i"),
+      '#tree' => TRUE,
+      '#description' => t('When requesting funds for additional Tripal development, 
+       it is important to report the breadth of of funding sources for Tripal sites.  
+       P0lease condiser sharing this information by providing the granting
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc]: simple typos

```suggestion
       Please consider sharing this information by providing the granting 
```

File: tripal/includes/tripal.registration.inc [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    
+  //Check for empty funding periods and remove them.
+  $j = 1;
+  foreach ($form_state['values']['funding'] as $funding_source) {
+    if (!empty($funding_source['tripal_reg_site_agency']) && !empty($funding_source['tripal_reg_site_grant'])) {
+      $form_state['values']['fundings'][$j] = $funding_source;
+      $j++;
+    }
+  }
+  $form_state['values']['funding'] = $form_state['values']['fundings'];
```
Comment (bradfordcondon) [inline tripal/includes/tripal.registration.inc]:
```
Notice: Undefined index: fundings in tripal_registration_form_submit() (line 262 of /Users/bc/tripal/sites/all/modules/tripal/tripal/includes/tripal.registration.inc).
Warning: Invalid argument supplied for foreach() in tripal_registration_remote_submit() (line 310 of /Users/bc/tripal/sites/all/modules/tripal/tripal/includes/tripal.registration.inc).
```

```suggestion
  $form_state['values']['funding'] = $form_state['values']['fundings'] ?? [];
```

**Review by bradfordcondon — CHANGES_REQUESTED** [review]
- I registered... i dont know that my submitted info should be publicly available at http://tripal.info/registration/tripal-test ?
- similarly, i think this form needs a data usage policy of some sort.
- If someone else has a site named tripal-test, is it going to overwrite my submitted info?
- I opted in, and then opted out.  the record on tripal.info is still there, still available to anonymous users.  should it/will it be deleted?

**Comment by almasaeed2010** [comment]
Looks good on my end as well. Thanks @shawnawsu

**Review by almasaeed2010 — APPROVED** [review]

**Comment by bradfordcondon** [comment]
I talked with Stephen, he said that the pages on tripal.info arent intended to be public.  If thats the case, then all we need to say is something along the lines of 

>Registration information will not be made public, and will be only used for grant-writing purposes. 

in the `Why Register your Site` markup

File: tripal/includes/tripal.registration.inc (hunk 1 above)
Comment (laceysanderson) [reply tripal/includes/tripal.registration.inc:41]: I can confirm this is still a problem. It looks like the PI email error no longer shows up but the site status still does.

File: tripal/includes/tripal.registration.inc (hunk 3 above)
Comment (laceysanderson) [reply tripal/includes/tripal.registration.inc]: I no longer see this error on testing :-)

File: tripal/includes/tripal.registration.inc (hunk 2 above)
Comment (laceysanderson) [reply tripal/includes/tripal.registration.inc]: This now appears checked by default 👍

File: tripal/includes/tripal.registration.inc (hunk 1 above)
Comment (shawnawsu) [reply tripal/includes/tripal.registration.inc:41]: I've pushed an update that should fix this problem, :)

File: tripal/includes/tripal.registration.inc (hunk 4 above)
Comment (shawnawsu) [reply tripal/includes/tripal.registration.inc]: I've collapsed these elements so I will resolve this.

File: tripal/includes/tripal.registration.inc (hunk 5 above)
Comment (shawnawsu) [reply tripal/includes/tripal.registration.inc]: Fixed

File: tripal/includes/tripal.registration.inc (hunk 6 above)
Comment (shawnawsu) [reply tripal/includes/tripal.registration.inc]: I've made the suggested change.

**Comment by shawnawsu** [comment]
I've made changes to the drupal 8 tripal.info module that handles the incoming data to address the possible overwriting of sites if they're named the same thing. Now the title of the node will be based on the site name and the site url. 

I've also installed the node view permission module so that only specific authenticated users can access the registrations. 

I've tested a submission and everything appears to be working but another round of testing would be greatly appreciated.

File: tripal/includes/tripal.registration.inc (hunk 1 above)
Comment (bradfordcondon) [reply tripal/includes/tripal.registration.inc:41]: looks good, great

**Review by bradfordcondon — APPROVED** [review]
- [x] single button opt-out
- [x] information not accessible to non-admins on tripal.info

etc etc .   Thanks for the changes looks great.

**Review by laceysanderson — CHANGES_REQUESTED** [review]
Hmmmm.... now when I opt out, it still creates a record on tripal.info

File: tripal/includes/tripal.registration.inc [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      $form_state['values']['fundings'][$j] = $funding_source;
+      $j++;
+    }
+  }
+  $form_state['values']['funding'] = $form_state['values']['fundings'] ?? [];
+  $registration = serialize($form_state);
+  variable_set('tripal_site_registration', $registration);
+
+  //Now send the updated info to the Tripal Site.
+  tripal_registration_remote_submit($registration);
```
Comment (laceysanderson) [inline tripal/includes/tripal.registration.inc]: This following should fix the problem I mentioned, ensuring that the site is only registered if they did not opt. out.

```suggestion
  // Only register with tripal.info if the user has not opt'd out.
  if ($form_state['values']['disable_tripal_reporting'] == FALSE) {
    tripal_registration_remote_submit($registration);
    drupal_set_message(t('Registration sent to tripal.info'), 'status');
  }
  else {
    drupal_set_message(t('You are not registered with tripal.info'), 'status');
  }
```

**Comment by shawnawsu** [comment]
@laceysanderson @spficklin I think this is good to go? I did additional testing and made a small tweak to the status message when disabling reporting.

**Review by laceysanderson — COMMENTED** [review]
Now when I opt out, it doesn't send info to tripal.info 🎉 

However, I can't delete the record on tripal.info any more (used to work)... Specifically, go to the `CANNOT DELETE` record on tripal.info and try to delete it. You will see
```
The website encountered an unexpected error. Please try again later.</br></br><em class="placeholder">Drupal\Core\Entity\EntityStorageException</em>: SQLSTATE[42S02]: Base table or view not found: 1146 Table &#039;tripalin_tripal_drupal8.field_collection_item__field_version&#039; doesn&#039;t exist: DELETE FROM {field_collection_item__field_version} 
WHERE entity_id = :db_condition_placeholder_0; Array
(
    [:db_condition_placeholder_0] =&gt; 196
)
 in <em class="placeholder">Drupal\Core\Entity\Sql\SqlContentEntityStorage-&gt;delete()</em> (line <em class="placeholder">795</em> of <em class="placeholder">core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php</em>).
```

**Comment by shawnawsu** [comment]
So the delete problem is down to entity/field definitions being out of date (http://tripal.info/admin/reports/status#error). At some point, someone updated the site and the process wasn't done correctly and now config and db are out of sync. It looks like someone upgraded core to 8.7.2 (which came out May 23) and that is probably when it happened. Now a drush entup (https://drushcommands.com/drush-8x/core/entity-updates/) needs to be run. Since we don't really have a dev environment for this site I'm leary to do it.

**Review by laceysanderson — APPROVED** [review]
Ok, since that is independent of this PR, I believe we are ready to merge! 🎉

**Pull request merged**